Environment_Engine/Compute/OpeningsFromGlazingRatio.cs

```diff
@@ -0,0 +1,98 @@
+        [Input("glazingRatio", "Ratio of total external panel surface area and glazed area. Ratio as decimal {0-1}.")]
+        [Input("panelsToIgnore", "Optional input for selecting a collection of Environment Panels to be ignored in the calculation")]
+        [Input("sillHeight", "Optional input for defining the distance between the base of the panel and the bottom of the opening - default: 0.5m.")]
+        [Input("openingHeight", "Optional input for defining the height of opening - default: 1.2m.")]
+        [Input("openingType", "Optional input for defining the opening type of the output Openings, can be either 'Glazing' or 'Door' - default: Glazing.")]
+        [Output("openings", "Returns the openings.")]
+        public static List<Opening> OpeningsFromGlazingRatio(
+            List<List<Panel>> panelsAsSpaces,
+            List<Point> glazingLocations,
+            double glazingRatio,
+            List<Panel> panelsToIgnore = null,
+            double sillHeight = 0.5,
+            double openingHeight = 1.2,
+            OpeningType openingType = OpeningType.Glazing
+            ) 
+        {
+            switch (openingType)
+            {
+                case OpeningType.Undefined:
+                case OpeningType.Door:
+                case OpeningType.Glazing:
+                    break;
+                default:
+                    BH.Engine.Base.Compute.RecordError($"Only openings of type 'Glazing' and 'Door' are supported.");
+                    return new List<Opening>();
+            }
+
+            List<Panel> wallPanels = panelsAsSpaces.SelectMany(x => x).Distinct().ToList();
+            List<Panel> externalPanels = wallPanels.IsExternal().Item1;
+            List<Panel> filteredPanels = externalPanels.RemovePanels(panelsToIgnore ?? new List<Panel>());
+
+            double totalArea = filteredPanels.Select(x=>x.Area()).Sum();
+            double existingOpeningArea = filteredPanels.Select(x => x.Openings.Select(y => y.Polyline().Area()).Sum()).Sum();
+            double windowWidth = WindowWidth(totalArea, glazingRatio, glazingLocations.Count(), openingHeight, existingOpeningArea);
+
+            OpeningOption option = new OpeningOption()
+            {
+                Height = openingHeight,
+                Width = windowWidth,
+                SillHeight = sillHeight,
+                Type = openingType,
+            };
+
+            List<Opening> openings = new List<Opening>();
+
+            foreach(Point point in glazingLocations)
+            {
+                Opening opening = Create.Opening(point, option, filteredPanels);
+
+                if(opening != null)
+                {
+                    openings.Add(opening);
+                }
+                else
+                {
+                    BH.Engine.Base.Compute.RecordWarning($"Could not find a host panel for the opening at ({point.X}, {point.Y}, {point.Z}).");
+                }
+            }
+
+            return openings;
+        }
+
```

Comment:
Please put in method separators and ideally the separators for `Private methods`. Though I'm unsure if this is needed as a private method? Can you give a bit more detail on this design choice?